### PR TITLE
Adjust oauth flow to use the middleman

### PR DESCRIPTION
**Pull request by TalMalka123**

### CODE REVIEW

File: dagshub/auth/oauth.py [hunk 1]
```diff
@@ -1,48 +1,52 @@
...
 import requests
 import urllib
 import uuid
 from dagshub.common import config
+import concurrent.futures
+# import getpass
+# import sys
+# import pytimedinput
 
+# CODE_INPUT_TIMEOUT = 60
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:12]: Remove dead code
```suggestion
```

File: dagshub/auth/oauth.py [hunk 2]
```diff
@@ -1,48 +1,52 @@
...
+    middle_man_uuid = uuid.uuid4()
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        future = executor.submit(call_middleman, dagshub_url, middle_man_uuid)
+        link_prompt = f"Go to {dagshub_url}/authorize?state={state}&client_id={client_id}" \
+                      f"&middleman_uuid={middle_man_uuid} to authorize the client."
         print(link_prompt)
-        code, timed_out = pytimedinput.timedInput(
-            prompt=code_prompt, timeout=code_input_timeout
+        res = future.result()
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:40]: There's no need for multithreading here.
1. Print the prompt
2. Do the POST

File: dagshub/auth/oauth.py [hunk 3]
```diff
@@ -1,48 +1,52 @@
...
+
+    if res.status_code != 200:
+        # maybe add a different error for the timeout flow
+        raise Exception(
+            f"Error while getting OAuth code: HTTP {res.status_code}, Body: {res.json()}"
         )
-        if timed_out:
-            raise RuntimeError("Timed out input of OAuth code")
+    code = res.json()["Code"]
+    print(code)
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:48]: Don't print the code

File: dagshub/auth/oauth.py [hunk 4]
```diff
@@ -52,9 +56,53 @@ def oauth_flow(
             f"Error while getting OAuth token: HTTP {res.status_code}, Body: {res.json()}"
         )
     token = res.json()
+
     logger.debug(f"Got token: {token}")
+    # Guy- should we print so massage to let the user know he is now logged in?
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:61]: That's what the logger.debug line does - don't add anything on top of it, that's just polluting the user's output

File: dagshub/auth/oauth.py [hunk 5]
```diff
@@ -52,9 +56,53 @@ def oauth_flow(
...
+#         data={"client_id": client_id, "code": code, "state": state},
+#     )
+#     if res.status_code != 200:
+#         raise Exception(
+#             f"Error while getting OAuth token: HTTP {res.status_code}, Body: {res.json()}"
+#         )
+#     token = res.json()
+#     logger.debug(f"Got token: {token}")
+#     return token
+
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:105]: Definitely delete dead code - if you ever want it again, it exists in git history
```suggestion
```

**Review by guysmoilov — CHANGES_REQUESTED** [review]

File: dagshub/auth/oauth.py [hunk 6]
```diff
@@ -32,20 +19,20 @@ def oauth_flow(
...
-        print(link_prompt)
-        res = future.result()
+    link_prompt = f"Go to {dagshub_url}/authorize?state={state}&client_id={client_id}" \
+                  f"&middleman_uuid={middle_man_uuid} to authorize the client."
+    print(link_prompt)
+
+    res = requests.post(
+        f"{dagshub_url}/middleman",
+        data={"uuid": middle_man_uuid},
+    )
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:29]: Check what it looks like when you get a timeout

**Review by guysmoilov — APPROVED** [review]

File: dagshub/auth/oauth.py (hunk 6 above)
Comment (TalMalka123) [reply dagshub/auth/oauth.py:29]: image
its clear

**Comment by guysmoilov** [comment]
Wait with merging this until the backend supports the new flow and we test it out fully

File: dagshub/auth/oauth.py [hunk 7]
```diff
@@ -17,27 +19,28 @@ def oauth_flow(
...
             f"Error while getting OAuth code: HTTP {res.status_code}, Body: {res.json()}"
         )
-    code = res.json()["Code"]
+    code = res.json()
 
-    res = requests.post(
+    res = httpx.post(
         f"{dagshub_url}/access_token",
-        data={"client_id": client_id, "code": code, "state": state},
+        data={"client_id": client_id, "code": code, "state": state}, timeout=None
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:41]: This request shouldn't take long
```suggestion
        data={"client_id": client_id, "code": code, "state": state}
```

File: dagshub/auth/oauth.py [hunk 8]
```diff
@@ -1,57 +1,52 @@
...
-    code_prompt = "Code:"
-    if code_input_timeout <= 0:
-        print(link_prompt)
-        code = getpass.getpass(code_prompt)
-    else:
-        if not sys.__stdin__.isatty():
-            raise OauthNonInteractiveShellException(
-                "Can't perform OAuth in a non-interactive shell. "
-                "Please get a token using this command in a shell: dagshub login"
-            )
```
Comment (guysmoilov) [inline dagshub/auth/oauth.py:38]: You still need this check. If the user has no token, and the shell isn't interactive, it should quit right way.

**Review by guysmoilov — CHANGES_REQUESTED** [review]

**Pull request merged**